lib/src/patch.dart

```diff
@@ -0,0 +1,57 @@
+// Copyright 2018 Google LLC
+//
+// Use of this source code is governed by an MIT-style
+// license that can be found in the LICENSE file or at
+// https://opensource.org/licenses/MIT.
+
+import 'package:source_span/source_span.dart';
+
+class Patch {
+  /// Selection to be replaced
+  final FileSpan selection;
+
+  /// Text to replace the selection with.
+  final String replacement;
+
+  final PatchType type;
+
+  /// Constructs a normal patch that replaces [selection] with [replacement].
+  const Patch(this.selection, this.replacement) : type = PatchType.normal;
+
+  const Patch.prepend(this.replacement)
+      : type = PatchType.prepend,
+        selection = null;
+
+  const Patch.append(this.replacement)
+      : type = PatchType.append,
+        selection = null;
+
+  /// Applies a series of non-overlapping patches to the text of a file.
+  static String applyAll(SourceFile file, List<Patch> patches) {
+    var normalPatches =
+        patches.where((p) => p.type == PatchType.normal).toList();
+    normalPatches.sort((a, b) => a.selection.compareTo(b.selection));
+    var buffer = StringBuffer();
+    int offset = 0;
+    patches
+        .where((p) => p.type == PatchType.prepend)
+        .map((p) => p.replacement)
+        .forEach(buffer.write);
+    for (var patch in normalPatches) {
+      if (patch.selection.start.offset < offset) {
+        throw new Exception("Can't apply overlapping patches.");
```

Comment:
```suggestion
        throw ArgumentError("Can't apply overlapping patches.");
```

If this ever actually gets thrown, it's going to be very painful to figure out why. Consider including [highlights](https://pub.dartlang.org/documentation/source_span/latest/source_span/SourceSpan/highlight.html) of the overlapping patches to indicate where they are.